client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,318 @@
+//! This module enables the tracking of workers that are associated with a client instance.
+//! This is needed to implement Eager Workflow Start, a latency optimization in which the client,
+//!  after reserving a slot, directly forwards a WFT to a local worker.
+
+use parking_lot::RwLock;
+use std::collections::{hash_map::Entry::Vacant, HashMap};
+
+use temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse;
+
+/// This trait is implemented by an object associated with a worker, which provides WFT processing slots.
+#[cfg_attr(test, mockall::automock)]
+pub trait SlotProvider: std::fmt::Debug {
+    /// A unique identifier for the worker.
+    fn id(&self) -> &str;
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> &str;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> &str;
+    /// Try to reserve a slot on this worker.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot + Send>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(
+        self: Box<Self>,
+        task: PollWorkflowTaskQueueResponse,
+    ) -> Result<(), anyhow::Error>;
+}
+
+/// This trait enables local workers to made themselves visible to a shared client instance.
+/// There can only be one worker registered per namespace+queue_name+client, others will get ignored.
+#[cfg_attr(test, mockall::automock)]
+pub trait WorkerRegistry {
+    /// Register a local worker that can provide WFT processing slots.
+    fn register(&self, provider: Box<dyn SlotProvider + Send + Sync>);
+    /// Unregister a provider, typically when its worker starts shutdown.
+    fn unregister(&self, id: &str);
+}
+
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+struct SlotKey {
+    namespace: String,
+    task_queue: String,
+}
+
+impl SlotKey {
+    fn new(namespace: String, task_queue: String) -> SlotKey {
+        SlotKey {
+            namespace,
+            task_queue,
+        }
+    }
+}
+
+/// Implements a [WorkerRegistry] and provides a convenient method
+/// to find compatible slots within the collection.
+/// This is an inner class to hide the mutex.
+#[derive(Default, Debug)]
+struct SlotManagerImpl {
+    /// Maps keys, i.e., namespace#task_queue, to provider.
+    providers: HashMap<SlotKey, Box<dyn SlotProvider + Send + Sync>>,
+    /// Maps ids to keys in `providers`.
+    index: HashMap<String, SlotKey>,
```

Comment:
We don't need to use UUIDs here, the easiest way to do this would be to use `SlotMap` as seen here: https://github.com/temporalio/sdk-core/blob/ee7f6caaa9ad00bca2e2db8093d444ce563d6016/core/src/worker/workflow/machines/workflow_machines.rs#L129

When registering, workers get back a key, and they can hold onto it to unregister themselves.

Reply:
I would have the slot map map worker keys to providers, and then have the extra map for `SlotKey -> WorkerKey`

Reply:
This is related to the discussion above of whether we want the registry to identify the workers with a stable id, for debugging, or other future uses... But if both of you think there is no value, I'm happy to get rid of uuids and switch to registry-generated keys...

Reply:
Yeah, I don't think there's any reason to use uuids.

Reply:
Done, uuids replaced by SlotMap keys.